[AVSampleBufferVideoRenderer expectMinimumUpcomingSampleBufferPresentationTime] could be called multiple times with the same value

jyavenard opened the pull request:
#### d23aa8496a633b7784ac6b812f4779360ac5243a
<pre>
[AVSampleBufferVideoRenderer expectMinimumUpcomingSampleBufferPresentationTime] could be called multiple times with the same value
<a href="https://bugs.webkit.org/show_bug.cgi?id=293333">https://bugs.webkit.org/show_bug.cgi?id=293333</a>
<a href="https://rdar.apple.com/151737336">rdar://151737336</a>

Reviewed by Youenn Fablet.

Also don&apos;t set the expectMinimumUpcomingSampleBufferPresentationTime while seeking and
add extra logging with the minimum upcoming time.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::hasIncomingOutOfOrderFrame const): Ignore samples that are marked as non-displaying.
(WebCore::VideoMediaSampleRenderer::minimumUpcomingSampleTime const): Same

Canonical link: <a href="https://commits.webkit.org/295305@main">https://commits.webkit.org/295305@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f2619ec0889d093894343e6699e835cf26fd12e8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104621 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24331 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14749 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109833 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55292 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106661 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24720 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32876 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79448 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107627 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19229 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94425 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59760 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19015 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12499 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54665 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88712 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12551 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112224 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31782 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23403 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88535 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32146 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90658 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88154 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22472 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33052 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10822 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/27099 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31709 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37051 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31502 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34840 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -599,8 +603,11 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
                 format = CVPixelBufferGetPixelFormatType(imageBuffer.get());
-            }
-            RELEASE_LOG_DEBUG(MediaPerformance, "VideoMediaSampleRenderer decoding rate:%0.1fHz rolling:%0.1f decoder rate:%0.1fHz compressed queue:%u decoded queue:%zu bframes:%d hw:%d format:%s", 1.0f / Seconds { now - std::exchange(m_timeSinceLastDecode, now) }.value(), m_frameRateMonitor.observedFrameRate(), 1.0f / Seconds { now - startTime }.value(), m_compressedSampleQueueSize.load(), decod ...
+                Ref mediaSample = MediaSampleAVFObjC::create((*result)[0].get(), 0);
```
It seems a bit wasteful to create a media sample for debug logging, no?

jyavenard replied:
> It seems a bit wasteful to create a media sample for debug logging, no?

this code is conditional, and is only called if the log level for MediaPerformance is set to debug, it will have zero impact on performance otherwise.

jyavenard replied:
I read the value from CMSampleBuffer directly

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -599,8 +610,10 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
+                auto presentationTime = PAL::CMSampleBufferGetOutputPresentationTimeStamp((*result)[0].get());
+                RELEASE_LOG_DEBUG(MediaPerformance, "VideoMediaSampleRenderer pts:%0.2f minimum upcoming:%0.2f decoding rate:%0.1fHz rolling:%0.1f decoder rate:%0.1fHz compressed queue:%u decoded queue:%zu bframes:%d hw:%d format:%s", PAL::CMTimeGetSeconds(presentationTime), m_lastMinimumUpcomingPresentationTime.toDouble(), 1.0f / Seconds { now - std::exchange(m_timeSinceLastDecode, now) }.value( ...
+            } else
+                RELEASE_LOG_DEBUG(MediaPerformance, "VideoMediaSampleRenderer pts:--- minimum upcoming:%0.2f decoding rate:%0.1fHz rolling:%0.1f decoder rate:%0.1fHz compressed queue:%u decoded queue:%zu bframes:%d hw:%d format:%s", m_lastMinimumUpcomingPresentationTime.toDouble(), 1.0f / Seconds { now - std::exchange(m_timeSinceLastDecode, now) }.value(), m_frameRateMonitor.observedFrameRate(),  ...
```
Can we merge both by having variables with a default value and overrides in `if (result->size() && (*result)[0]) {` statement?

jyavenard replied:
I will remove all this code in https://webkit.org/b/293422

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -254,34 +256,37 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
         if (flushId != m_flushId)
             return false;
-        if (sample->presentationTime() < time)
+        if (!sample->isNonDisplaying() && sample->presentationTime() < time)
```
We are starting to handle non displaying samples, which is fine. But the commit log could be augmented to describe the changes more precisely.

youennf commented on Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
```diff
@@ -499,6 +504,9 @@ static OSStatus sampleCallback(CMBufferRef buffer, void* refcon)
     if (m_compressedSampleQueue.isEmpty())
         return;
 
+    if (m_hasOutOfOrderFrames && m_compressedSampleQueue.size() < MaximumSlidingWindowLength && m_compressedSampleQueueSize >= MaximumSlidingWindowLength)
```
This is not easy to read, maybe we should rename `m_compressedSampleQueueSize` to `m_compressedSamplesCount` maybe.

jyavenard committed:
[AVSampleBufferVideoRenderer expectMinimumUpcomingSampleBufferPresentationTime] could be called multiple times with the same value
https://bugs.webkit.org/show_bug.cgi?id=293333
rdar://151737336

Reviewed by Youenn Fablet.

Also don't set the expectMinimumUpcomingSampleBufferPresentationTime while seeking and
add extra logging with the minimum upcoming time.

* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.mm:
(WebCore::VideoMediaSampleRenderer::hasIncomingOutOfOrderFrame const): Ignore samples that are marked as non-displaying.
(WebCore::VideoMediaSampleRenderer::minimumUpcomingSampleTime const): Same

Canonical link: https://commits.webkit.org/295305@main
